include/util/packed_vector.hpp

```diff
@@ -31,188 +35,353 @@ inline void write(storage::io::FileWriter &writer,
 
 namespace detail
 {
+
+template <typename WordT, typename T>
+inline T get_lower_half_value(WordT word,
+                              WordT mask,
+                              std::uint8_t offset,
+                              typename std::enable_if_t<std::is_integral<T>::value> * = 0)
+{
+    return static_cast<T>((word & mask) >> offset);
+}
+
+template <typename WordT, typename T>
+inline T
+get_lower_half_value(WordT word, WordT mask, std::uint8_t offset, typename T::value_type * = 0)
+{
+    return T{static_cast<typename T::value_type>((word & mask) >> offset)};
+}
+
+template <typename WordT, typename T>
+inline T get_upper_half_value(WordT word,
+                              WordT mask,
+                              std::uint8_t offset,
+                              typename std::enable_if_t<std::is_integral<T>::value> * = 0)
+{
+    return static_cast<T>((word & mask) << offset);
+}
+
+template <typename WordT, typename T>
+inline T
+get_upper_half_value(WordT word, WordT mask, std::uint8_t offset, typename T::value_type * = 0)
+{
+    return T{static_cast<typename T::value_type>((word & mask) << offset)};
+}
+
+template <typename WordT, typename T>
+inline WordT set_lower_value(WordT word, WordT mask, std::uint8_t offset, T value)
+{
+    return (word & ~mask) | (static_cast<WordT>(value) << offset);
+}
+
+template <typename WordT, typename T>
+inline WordT set_upper_value(WordT word, WordT mask, std::uint8_t offset, T value)
+{
+    return (word & ~mask) | (static_cast<WordT>(value) >> offset);
```

Comment:
I'm going to add a static assertion to check that `WordT` is an unsigned type, if that works for you.